### PR TITLE
T12729

### PR DESCRIPTION
**Pull request by smspillaz**
Obviously, there are a lot of commits here which are now redundant. I'd suggest just having a look at the overall diff and we can squash as appropriate

https://phabricator.endlessm.com/T12729

### CODE REVIEW

File: js/misc/missionGameService.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+        let name = 'com.endlessm.Showmehow.Service';
+        let path = '/com/endlessm/Showmehow/Service';
+
+        /* Connect to the service and refresh the content once we have a connection */
+        Showmehow.ServiceProxy.new_for_bus(Gio.BusType.SESSION, 0, name, path, null,
+                                           Lang.bind(this, function(source, result) {
+            try {
+                this._service = Showmehow.ServiceProxy.new_for_bus_finish(result);
+            } catch (e) {
+                logError(e, "Error occurred in creating ShowmehowServiceProxy");
```
Comment (cosimoc) [inline js/misc/missionGameService.js]: Single quotes, unless the string is marked for translation.

File: js/misc/missionGameService.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+        }));
+    },
+
+    noteEventOccurrence: function(event) {
+        this._service.call_lesson_event(event, null, null);
+    }
+});
+Signals.addSignalMethods(MissionChatboxTextService.prototype);
+
+function getService() {
```
Comment (cosimoc) [inline js/misc/missionGameService.js]: If you have this here, you don't really need to set anything on `global`, do you?

File: js/ui/main.js [hunk 3]
```diff
@@ -65,6 +65,7 @@ let shellDBusService = null;
 let shellMountOpDBusService = null;
 let screenSaverDBus = null;
 let socialBar = null;
+let missionChatbox = null;
```
Comment (cosimoc) [inline js/ui/main.js]: Looks like this is never assigned?

File: js/ui/panel.js [hunk 4]
```diff
@@ -196,7 +196,7 @@ const VariableSpeedAnimation = new Lang.Class({
 
             return false;
         }
-        
+
```
Comment (cosimoc) [inline js/ui/panel.js]: Spurious change

File: js/ui/sideComponent.js [hunk 5]
```diff
@@ -14,7 +14,7 @@ const SIDE_COMPONENT_ROLE = 'eos-side-component';
 /**
  * isSideComponentWindow:
  * @metaWindow: an instance of #Meta.Window
- * @return: whether the #Meta.Window belongs to a #SideComponent 
+ * @return: whether the #Meta.Window belongs to a #SideComponent
```
Comment (cosimoc) [inline js/ui/sideComponent.js]: Spurious change

File: js/js-resources.gresource.xml [hunk 6]
```diff
@@ -116,5 +118,6 @@
     <file>ui/components/polkitAgent.js</file>
     <file>ui/components/socialBar.js</file>
     <file>ui/components/updaterManager.js</file>
+    <file>ui/components/missionChatbox.js</file>
```
Comment (cosimoc) [inline js/js-resources.gresource.xml]: This file is missing

File: js/ui/windowManager.js [hunk 7]
```diff
@@ -756,6 +758,26 @@ const WindowManager = new Lang.Class({
...
+        this._missionGameService = MissionGameService.getService();
+
+        /* For mission we add another event handler for grab-op-begin
+         * to see when a window has moved or is moving.
+         *
+         * This is expensive, so we only do it if we are interested
+         * in window-move events */
+        this._missionGameService.connect('listening-for-events', Lang.bind(this, function(service, events) {
+            if (events.indexOf('window-moved') !== -1) {
+                this._missionGrabOpBeginHandler = global.display.connect('grab-op-begin',
```
Comment (cosimoc) [inline js/ui/windowManager.js]: This could be a local variable

File: js/misc/missionGameService.js [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+        this.parent();
+
+        /* Null-initialise service for now, but we'll set it later */
+        this._service = null;
+
+        let name = 'com.endlessm.Showmehow.Service';
+        let path = '/com/endlessm/Showmehow/Service';
+
+        /* Connect to the service and refresh the content once we have a connection */
+        Showmehow.ServiceProxy.new_for_bus(Gio.BusType.SESSION, 0, name, path, null,
```
Comment (smspillaz) [inline js/misc/missionGameService.js]: Yeah, if we can do that, it would simplify a number of things. Will do.

File: js/ui/main.js (hunk 3 above)
Comment (smspillaz) [inline js/ui/main.js]: Hmm, it should be. I'll need to check that.

File: js/misc/missionGameService.js (hunk 2 above)
Comment (smspillaz) [inline js/misc/missionGameService.js]: I suppose not. I'll remove the things set on `global`. Though I suppose that raises the question - what is the preferred way of handling singletons like these? `global` or having a singleton getter?

File: js/js-resources.gresource.xml (hunk 6 above)
Comment (smspillaz) [inline js/js-resources.gresource.xml]: oh, how awkward. I'll git add it now!

File: js/misc/missionGameService.js [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+    _init: function() {
+        let name = 'com.endlessm.Showmehow.Service';
+        let path = '/com/endlessm/Showmehow/Service';
+
+        /* Connect to the service and refresh the content once we have a connection */
+        try {
+            this._service = Showmehow.ServiceProxy.new_for_bus_sync(Gio.BusType.SESSION,
+                                                                    0, name, path, null);
+        } catch (e) {
+            logError(e, "Error occurred in creating ShowmehowServiceProxy");
```
Comment (cosimoc) [inline js/misc/missionGameService.js]: This should return here.
Also, single quotes.

File: js/misc/missionGameService.js [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+         * of tuples (which contain strings). */
+        this._service.connect('listening-for-lesson-events', Lang.bind(this, function(proxy, interestingEvents) {
+            this.emit('listening-for-events', interestingEvents.deep_unpack().map(function(i) {
+                return i[0];
+            }));
+        }));
+    },
+
+    noteEventOccurrence: function(event) {
+        this._service.call_lesson_event(event, null, null);
```
Comment (cosimoc) [inline js/misc/missionGameService.js]: Should probably check for `this._service != null`

File: js/ui/sideComponent.js (hunk 5 above)
Comment (cosimoc) [inline js/ui/sideComponent.js]: Can you please remove this?

File: data/theme/gnome-shell.css [hunk 11]
```diff
@@ -1746,6 +1746,84 @@ StScrollBar StButton#vhandle:active {
     link-color: #ccccff;
 }
 
+/* Mission text */
```
Comment (cosimoc) [inline data/theme/gnome-shell.css]: Looks like all these changes are not needed anymore.

File: js/ui/main.js (hunk 3 above)
Comment (smspillaz) [inline js/ui/main.js]: Ah, yes it is in `missionChatbox.js` which wasn't added earlier.

File: js/ui/sideComponent.js (hunk 5 above)
Comment (smspillaz) [inline js/ui/sideComponent.js]: Ah - I separated those whitespace fixes out into separate commits. But I suppose we don't want them, so I dropped the commits entirely.

**Comment by cosimoc** [comment]
Obsoleted by #188

**Pull request closed without merging**